src/main/kotlin/org/koitharu/kotatsu/parsers/site/en/LineWebtoonsParser.kt

```diff
@@ -0,0 +1,426 @@
+	class English(context: MangaLoaderContext) : LineWebtoonsParser(context, MangaSource.LINEWEBTOONS_EN)
+	@MangaSourceParser("LINEWEBTOONS_ZH", "Line Webtoons Chinese", "zh", type = ContentType.MANGA)
+	class Chinese(context: MangaLoaderContext) : LineWebtoonsParser(context, MangaSource.LINEWEBTOONS_ZH)
+	@MangaSourceParser("LINEWEBTOONS_TH", "Line Webtoons Thai", "th", type = ContentType.MANGA)
+	class Thai(context: MangaLoaderContext) : LineWebtoonsParser(context, MangaSource.LINEWEBTOONS_TH)
+	@MangaSourceParser("LINEWEBTOONS_ID", "Line Webtoons Indonesian", "id", type = ContentType.MANGA)
+	class Indonesian(context: MangaLoaderContext) : LineWebtoonsParser(context, MangaSource.LINEWEBTOONS_ID)
+	@MangaSourceParser("LINEWEBTOONS_ES", "Line Webtoons Spanish", "es", type = ContentType.MANGA)
+	class Spanish(context: MangaLoaderContext) : LineWebtoonsParser(context, MangaSource.LINEWEBTOONS_ES)
+	@MangaSourceParser("LINEWEBTOONS_FR", "Line Webtoons French", "fr", type = ContentType.MANGA)
+	class French(context: MangaLoaderContext) : LineWebtoonsParser(context, MangaSource.LINEWEBTOONS_FR)
+	@MangaSourceParser("LINEWEBTOONS_DE", "Line Webtoons German", "de", type = ContentType.MANGA)
+	class German(context: MangaLoaderContext) : LineWebtoonsParser(context, MangaSource.LINEWEBTOONS_DE)
+
+}
+
+
+private class WebtoonsUrlSigner(val secret: String) {
+	private val mac = Mac.getInstance("HmacSHA1").apply {
+		this.init(SecretKeySpec(secret.encodeToByteArray(), "HmacSHA1"))
+	}
+
+	private fun getMessage(url: String, msgpad: String): String {
+		return url.substring(0, 0xFF.coerceAtMost(url.length)) + msgpad
+	}
+
+	private fun getMessageDigest(s: String): String {
+		var signedMessage: ByteArray
+		synchronized(mac) { signedMessage = mac.doFinal(s.toByteArray()) }
+
+		// we don't use the context.encodeBase64 here because it adds newlines and doesn't add padding
+		// we, however, need padding and no newlines
+		return encodeBase64(signedMessage)
+	}
+
+
+	fun makeEncryptUrl(s: String): String {
+		return makeEncryptUrlCore(
+			s,
+			java.lang.String.valueOf(Calendar.getInstance().timeInMillis),
+		)
+	}
+
+	private fun makeEncryptUrlCore(url: String, msgpad: String): String {
+		val digest = URLEncoder.encode(getMessageDigest(getMessage(url, msgpad)), "utf-8")
+		return url + if (url.contains("?")) {
+			"&"
+		} else {
+			"?"
+		} + "msgpad=${msgpad}&md=${digest}"
+	}
+}
+
+
+private val INT_TO_BASE64: CharArray = charArrayOf(
+	'A',
+	'B',
+	'C',
+	'D',
+	'E',
+	'F',
+	'G',
+	'H',
+	'I',
+	'J',
+	'K',
+	'L',
+	'M',
+	'N',
+	'O',
+	'P',
+	'Q',
+	'R',
+	'S',
+	'T',
+	'U',
+	'V',
+	'W',
+	'X',
+	'Y',
+	'Z',
+	'a',
+	'b',
+	'c',
+	'd',
+	'e',
+	'f',
+	'g',
+	'h',
+	'i',
+	'j',
+	'k',
+	'l',
+	'm',
+	'n',
+	'o',
+	'p',
+	'q',
+	'r',
+	's',
+	't',
+	'u',
+	'v',
+	'w',
+	'x',
+	'y',
+	'z',
+	'0',
+	'1',
+	'2',
+	'3',
+	'4',
+	'5',
+	'6',
+	'7',
+	'8',
+	'9',
+	'+',
+	'/',
+)
+
+private fun encodeBase64(arr: ByteArray): String {
```

Comment:
Try to use `encodeBase64` from the `MangaLoaderContext` class instead (if it possible) because it has another implementation on Android

Reply:
(I've left a comment regarding that):

```kotlin
// we don't use the context.encodeBase64 here because it adds newlines and doesn't add padding
// we, however, need padding and no newlines
```

That's why I am rolling out my own impl. For me to use it, `MangaLoaderContext.encodeBase64` would need accept options controlling that

Btw, wouldn't it be simpler to have a common implementation in `kotatsu-parsers`, not relying on system libs (AFAIU, the whole `MangaLoaderContext` thing is because `java.util.Base64` is not supported on low API levels).